relay-server/src/services/project.rs

```diff
@@ -579,52 +579,72 @@ impl Project {
         }
     }
 
-    /// Remove metric buckets that are not allowed to be ingested.
+    /// Remove metric buckets or tags that are not allowed to be ingested.
     fn filter_metrics(&self, metrics: &mut Vec<Bucket>) {
         let Some(state) = &self.state_value() else {
             return;
         };
 
-        Self::apply_metrics_deny_list(&state.config.metrics, metrics);
-        Self::filter_disabled_namespace(state, metrics);
-    }
-
-    fn apply_metrics_deny_list(deny_list: &ErrorBoundary<Metrics>, buckets: &mut Vec<Bucket>) {
-        let ErrorBoundary::Ok(metrics) = deny_list else {
-            return;
-        };
+        let metrics_config = &state.config.metrics;
 
-        buckets.retain(|bucket| {
-            if metrics.denied_names.is_match(&bucket.name) {
+        metrics.retain_mut(|bucket| {
+            if Self::metric_name_denied(metrics_config, bucket) {
                 relay_log::trace!(mri = bucket.name, "dropping metrics due to block list");
                 false
+            } else if Self::metric_namespace_invalid(state, bucket) {
+                false
             } else {
+                Self::remove_tags(metrics_config, bucket);
                 true
             }
         });
     }
 
-    fn filter_disabled_namespace(state: &ProjectState, metrics: &mut Vec<Bucket>) {
-        metrics.retain(|metric| {
-            let Ok(mri) = MetricResourceIdentifier::parse(&metric.name) else {
-                relay_log::trace!(mri = metric.name, "dropping metrics with invalid MRI");
-                return false;
-            };
+    /// Removes tags based on user configured deny list.
+    fn remove_tags(deny_list: &ErrorBoundary<Metrics>, bucket: &mut Bucket) {
+        let ErrorBoundary::Ok(metrics) = deny_list else {
+            return;
+        };
 
-            let verdict = match mri.namespace {
-                MetricNamespace::Sessions => true,
-                MetricNamespace::Transactions => true,
-                MetricNamespace::Spans => state.has_feature(Feature::SpanMetricsExtraction),
-                MetricNamespace::Custom => state.has_feature(Feature::CustomMetrics),
-                MetricNamespace::Unsupported => false,
-            };
+        for tag_block_config in &metrics.denied_tags {
+            let name_pattern = &tag_block_config.name;
+            let tag_pattern = &tag_block_config.tag;
 
-            if !verdict {
-                relay_log::trace!(mri = metric.name, "dropping metric in disabled namespace");
+            if name_pattern.is_match(&bucket.name) {
+                bucket
+                    .tags
+                    .retain(|tag_key, _| !tag_pattern.is_match(tag_key));
             }
+        }
+    }
 
-            verdict
-        });
+    fn metric_name_denied(deny_list: &ErrorBoundary<Metrics>, bucket: &Bucket) -> bool {
+        let ErrorBoundary::Ok(metrics) = deny_list else {
```

Comment:
We now check the `ErrorBoundary` twice in two consecutive functions. If we pull this to the upper level, we could check it once, and then conditionally call the inner functions (or inline them, since we're really only talking about a single line of code, and a handful in the other case).